server/raft_test.go

```diff
@@ -844,3 +844,55 @@ func TestNRGNoResetOnAppendEntryResponse(t *testing.T) {
 	require_NotEqual(t, leader.pterm, 0)
 	require_NotEqual(t, leader.pindex, 0)
 }
+
+func TestNRGCandidateDontStepdownDueToLeaderOfPreviousTerm(t *testing.T) {
+	c := createJetStreamClusterExplicit(t, "R3S", 3)
+	defer c.shutdown()
+	c.waitOnLeader()
+
+	nc, _ := jsClientConnect(t, c.leader(), nats.UserInfo("admin", "s3cr3t!"))
+	defer nc.Close()
+
+	rg := c.createRaftGroup("TEST", 3, newStateAdder)
+	rg.waitOnLeader()
+
+	// create a candidate that has received entries while they were a follower in a previous term
+	candidate := rg.nonLeader().node().(*raft)
+	candidate.Lock()
+	candidate.switchState(Candidate)
+	candidate.pterm = 50
+	candidate.pindex = 70
+	candidate.term = 100
+	candidate.Unlock()
+
+	// leader term is behind candidate
+	leader := rg.leader().node().(*raft)
+	leader.Lock()
+	leader.term = candidate.pterm
+	leader.pterm = candidate.pterm
+	leader.pindex = candidate.pindex
+	leader.Unlock()
+
+	// Subscribe to the append entry subject.
+	sub, err := nc.SubscribeSync(leader.asubj)
+	require_NoError(t, err)
+
+	// Get the first append entry that we receive, should be heartbeat from leader of prev term
+	msg, err := sub.NextMsg(5 * time.Second)
+	require_NoError(t, err)
+
+	// Decode the append entry
+	ae, err := leader.decodeAppendEntry(msg.Data, nil, msg.Reply)
+	require_NoError(t, err)
+
+	// check that the append entry is from the leader of the previous term
+	require_Equal(t, ae.term, leader.term)
```

Comment:
Looks like we'll need to take the `leader` lock here:
```
==================
WARNING: DATA RACE
Write at 0x00c00183e168 by goroutine 104774:
  github.com/nats-io/nats-server/v2/server.(*raft).processAppendEntryResponse()
      /home/travis/gopath/src/github.com/nats-io/nats-server/server/raft.go:3521 +0x105
  github.com/nats-io/nats-server/v2/server.(*raft).runAsLeader()
      /home/travis/gopath/src/github.com/nats-io/nats-server/server/raft.go:2393 +0x10fe
  github.com/nats-io/nats-server/v2/server.(*raft).run()
      /home/travis/gopath/src/github.com/nats-io/nats-server/server/raft.go:1874 +0x4d8
  github.com/nats-io/nats-server/v2/server.(*raft).run-fm()
      <autogenerated>:1 +0x33
  github.com/nats-io/nats-server/v2/server.(*Server).startGoRoutine.func1()
      /home/travis/gopath/src/github.com/nats-io/nats-server/server/server.go:3824 +0x59
Previous read at 0x00c00183e168 by goroutine 104406:
  github.com/nats-io/nats-server/v2/server.TestNRGCandidateDontStepdownDueToLeaderOfPreviousTerm()
      /home/travis/gopath/src/github.com/nats-io/nats-server/server/raft_test.go:889 +0x6c5
  testing.tRunner()
      /home/travis/.gimme/versions/go1.22.5.linux.amd64/src/testing/testing.go:1689 +0x21e
  testing.(*T).Run.gowrap1()
      /home/travis/.gimme/versions/go1.22.5.linux.amd64/src/testing/testing.go:1742 +0x44
Goroutine 104774 (running) created at:
  github.com/nats-io/nats-server/v2/server.(*Server).startGoRoutine()
      /home/travis/gopath/src/github.com/nats-io/nats-server/server/server.go:3822 +0x1ba
  github.com/nats-io/nats-server/v2/server.(*Server).startRaftNode()
      /home/travis/gopath/src/github.com/nats-io/nats-server/server/raft.go:527 +0x20b8
  github.com/nats-io/nats-server/v2/server.(*cluster).createRaftGroupWithPeers()
      /home/travis/gopath/src/github.com/nats-io/nats-server/server/raft_helpers_test.go:148 +0x98d
  github.com/nats-io/nats-server/v2/server.(*cluster).createRaftGroupEx()
      /home/travis/gopath/src/github.com/nats-io/nats-server/server/raft_helpers_test.go:117 +0x316
  github.com/nats-io/nats-server/v2/server.(*cluster).createRaftGroup()
      /home/travis/gopath/src/github.com/nats-io/nats-server/server/raft_helpers_test.go:103 +0x324
  github.com/nats-io/nats-server/v2/server.TestNRGCandidateDontStepdownDueToLeaderOfPreviousTerm()
      /home/travis/gopath/src/github.com/nats-io/nats-server/server/raft_test.go:856 +0x2f7
  testing.tRunner()
      /home/travis/.gimme/versions/go1.22.5.linux.amd64/src/testing/testing.go:1689 +0x21e
  testing.(*T).Run.gowrap1()
      /home/travis/.gimme/versions/go1.22.5.linux.amd64/src/testing/testing.go:1742 +0x44
Goroutine 104406 (running) created at:
  testing.(*T).Run()
      /home/travis/.gimme/versions/go1.22.5.linux.amd64/src/testing/testing.go:1742 +0x825
  testing.runTests.func1()
      /home/travis/.gimme/versions/go1.22.5.linux.amd64/src/testing/testing.go:2161 +0x85
  testing.tRunner()
      /home/travis/.gimme/versions/go1.22.5.linux.amd64/src/testing/testing.go:1689 +0x21e
  testing.runTests()
      /home/travis/.gimme/versions/go1.22.5.linux.amd64/src/testing/testing.go:2159 +0x8be
  testing.(*M).Run()
      /home/travis/.gimme/versions/go1.22.5.linux.amd64/src/testing/testing.go:2027 +0xf17
  github.com/nats-io/nats-server/v2/server.TestMain()
      /home/travis/gopath/src/github.com/nats-io/nats-server/server/sublist_test.go:1602 +0x416
  main.main()
      _testmain.go:2329 +0x2d4
==================
    testing.go:1398: race detected during execution of test
```

Reply:
Just pushed to fix

Reply:
Done